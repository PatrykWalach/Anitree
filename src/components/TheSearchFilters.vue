<template>
  <v-dialog v-model="dialog" max-width="360px" scrollable>
    <v-card>
      <v-card-title>
        Filters
      </v-card-title>
      <v-card-text class="pa-0">
        <v-card-text>
          <v-row>
            <v-col cols="12">
              <span class="subheading">Type</span>

              <v-chip-group
                :value="form.type"
                @change="
                  type => {
                    form = { type }
                  }
                "
                column
                active-class="accent--text"
              >
                <v-chip
                  v-for="type in ['ANIME', 'MANGA']"
                  :key="type"
                  :value="type"
                  class="text-capitalize"
                >
                  {{ type.toLowerCase() }}
                </v-chip>
              </v-chip-group>
            </v-col>

            <v-col cols="12">
              <span class="subheading">Sort</span>

              <v-chip-group
                :value="form.sort"
                @change="
                  sort => {
                    form = { sort }
                  }
                "
                column
                active-class="accent--text"
              >
                <v-chip
                  v-for="{ value, text } in [
                    {
                      value: 'ID_DESC',
                      text: 'Date Added',
                    },
                    {
                      value: 'FAVOURITES_DESC',
                      text: 'Favorites',
                    },
                    {
                      value: 'POPULARITY_DESC',
                      text: 'Popularity',
                    },
                    {
                      value: 'START_DATE_DESC',
                      text: 'Release Date',
                    },
                    {
                      value: 'SCORE_DESC',
                      text: 'Score',
                    },
                    {
                      value: 'TITLE_ROMAJI',
                      text: 'Title',
                    },
                    {
                      value: 'TRENDING_DESC',
                      text: 'Trending',
                    },
                  ]"
                  :key="value"
                  :value="value"
                  class="text-capitalize"
                >
                  {{ text }}
                </v-chip>
              </v-chip-group>
            </v-col>

            <v-col cols="12">
              <span class="subheading">Status</span>
              <base-field
                :value="form.status"
                :before-transform="[
                  e => (e ? (e instanceof Array ? e : [e]) : []),
                ]"
                tag="v-chip-group"
                :after-transform="[
                  e => (e.length ? e : undefined),
                  status => {
                    form = { status }
                  },
                ]"
                multiple
                column
                active-class="accent--text"
              >
                <v-chip
                  v-for="status in [
                    'CANCELLED',
                    'FINISHED',
                    'NOT_YET_RELEASED',
                    'RELEASING',
                  ]"
                  :key="status"
                  filter
                  :value="status"
                  class="text-capitalize"
                >
                  {{
                    status
                      .toLowerCase()
                      .split(/_/)
                      .join(' ')
                  }}
                </v-chip>
              </base-field>
            </v-col>
          </v-row>
        </v-card-text>
        <v-list flat subheader>
          <v-subheader>Other</v-subheader>
          <v-list-item-group
            :value="form.onList"
            @change="
              onList => {
                form = { onList }
              }
            "
          >
            <v-list-item v-slot="{ active }" value="false">
              <v-list-item-content>
                <v-list-item-title>
                  Hide anime on my list
                </v-list-item-title>
              </v-list-item-content>
              <v-list-item-action>
                <v-checkbox v-model="active"></v-checkbox>
              </v-list-item-action>
            </v-list-item>
            <v-list-item v-slot="{ active }" value="true">
              <v-list-item-content>
                <v-list-item-title>
                  Only show anime on my list
                </v-list-item-title>
              </v-list-item-content>
              <v-list-item-action>
                <v-checkbox v-model="active"></v-checkbox>
              </v-list-item-action>
            </v-list-item>
          </v-list-item-group>

          <v-list-item-group
            :value="form.isAdult"
            @change="
              isAdult => {
                form = { isAdult }
              }
            "
          >
            <v-list-item v-slot="{ active }" value="true">
              <v-list-item-content>
                <v-list-item-title>
                  Adult
                </v-list-item-title>
              </v-list-item-content>
              <v-list-item-action>
                <v-checkbox v-model="active"></v-checkbox>
              </v-list-item-action>
            </v-list-item>
          </v-list-item-group>
        </v-list>
      </v-card-text>
      <v-divider></v-divider>
      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn text color="error" @click.stop="close">cancel</v-btn>
        <v-btn
          :disabled="!submitRequired"
          text
          color="primary"
          @click.stop="submit"
          >submit</v-btn
        >
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import { computed, createComponent } from '@vue/composition-api'

import { useDialog, useForm } from '@/mixins'
import BaseField from './BaseField.vue'

import { VChipGroup } from 'vuetify/lib'

export default createComponent({
  components: {
    BaseField,
    VChipGroup,
  },
  inheritAttrs: false,
  setup(_, { root }) {
    const { dialog } = useDialog(root)
    const { _form, query, form } = useForm(root)

    const close = () => {
      dialog.value = false
      _form.value = {}
    }

    const submit = () => {
      root.$router.replace({
        query: form.value,
      })
      close()
    }

    const submitRequired = computed(() => {
      const _query = query.value

      for (const [key, value] of Object.entries(_form.value)) {
        if (_query[key] !== value) {
          return true
        }
      }
      return false
    })

    return {
      _form,
      close,
      dialog,
      form,
      submit,
      submitRequired,
    }
  },
})
</script>
