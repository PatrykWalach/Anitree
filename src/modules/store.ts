import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
