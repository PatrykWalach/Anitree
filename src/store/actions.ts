export { titleActions as title } from './modules/title'

export { editActions as edit } from './modules/edit'
export { filterActions as filter } from './modules/filter'
export { shareActions as share } from './modules/share'
export { settingsActions as settings } from './modules/settings'
export { timelineActions as timeline } from './modules/timeline'
export { changesActions as changes } from './modules/changes'
