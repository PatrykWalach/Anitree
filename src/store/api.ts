import gql from 'graphql-tag'
import { FetchVariables } from './models'

import apollo from '../apollo'
import { AMedia } from '@/types'

export const query = gql`
  query($search: String, $idIn: [Int], $page: Int = 0) {
    Page(page: $page, perPage: 50) {
      media(search: $search, id_in: $idIn) {
        id
        bannerImage
        coverImage {
          extraLarge
          large
          medium
          color
        }
        format
        status
        tags {
          name
          id
          rank
        }
        siteUrl
        type
        studios(isMain: true) {
          nodes {
            name
            id
          }
        }
        startDate {
          year
          month
          day
        }
        endDate {
          year
          month
          day
        }
        title {
          romaji
          english
          native
        }
        description
        mediaListEntry {
          status
        }
        relations {
          edges {
            node {
              id
              type
            }
            id
            relationType
          }
        }
      }
    }
  }
`

export function fetchMediaApollo(variables: FetchVariables): Promise<AMedia[]> {
  return apollo
    .query<{ Page: { media: AMedia[] } }>({
      query,
      variables
    })
    .then(({ data }) => data.Page.media)
}
