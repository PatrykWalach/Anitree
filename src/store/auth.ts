import VueFunctionApi, { ref, Ref, computed } from 'vue-function-api'
import Vue from 'vue'
Vue.use(VueFunctionApi)

import { Token } from '@/types'
const stored = localStorage.getItem('AUTH')

const auth: Ref<Token | null> = ref((stored && JSON.parse(stored)) || null)

const token = computed(() => {
  if (
    auth.value
    // && Date.now() < parseInt(auth.value.expires_in)
  ) {
    return auth.value.access_token
  }
  return null
})

const CHANGE_TOKEN = async (payload: Token | null) => {
  if (payload) {
    const newAuth: Token = {
      ...payload,
      /* eslint-disable-next-line @typescript-eslint/camelcase */
      expires_in: (Date.now() + parseInt(payload.expires_in)).toString()
    }
    localStorage.setItem('AUTH', JSON.stringify(newAuth))
    auth.value = newAuth
  } else {
    auth.value = payload
  }
}

export default function useAuth() {
  return { auth, CHANGE_TOKEN, token, stored }
}
