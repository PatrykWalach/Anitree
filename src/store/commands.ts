import CompositionApi, { Ref, ref, watch } from '@vue/composition-api'
import {
  DeleteCommand,
  DeleteCommandConstructor,
} from './commands/DeleteCommand'
import { DeleteVariables, SaveVariables } from '@/graphql/schema/listEntry'
import { SaveCommand, SaveCommandConstructor } from './commands/SaveCommand'
import { Form } from '@/types'
import { FuzzyDate } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import Vue from 'vue'
import { useSettings } from './settings'

Vue.use(CompositionApi)

export interface Command {
  saveState?(): void | Promise<void>
  undo?(): void | Promise<void>
  execute(): void | Promise<void>
}

export interface ListCommand extends Command {
  backup: Form | null
  variables: SaveVariables | (DeleteVariables & Pick<SaveVariables, 'mediaId'>)
  done: boolean
  _class: 'SaveCommand' | 'DeleteCommand'
}

export const isDeleteCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is DeleteCommandConstructor & Pick<DeleteCommand, '_class'> =>
  e._class === 'DeleteCommand'

export const isSaveCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is SaveCommandConstructor & Pick<SaveCommand, '_class'> =>
  e._class === 'SaveCommand'

const stored = localStorage.getItem('CHANGES')

const history: Ref<ListCommand[]> = ref(
  (stored &&
    JSON.parse(stored, (key, value) => {
      if (value instanceof Object) {
        if (isSaveCommand(value)) {
          return new SaveCommand(value)
        }
        if (isDeleteCommand(value)) {
          return new DeleteCommand(value)
        }
      }
      return value
    })) ||
    [],
)

watch(() => {
  if (useSettings().cacheChanges.value) {
    localStorage.setItem('CHANGES', JSON.stringify(history.value))
  }
})
;(async () => {
  for (const command of history.value.filter(({ done }) => !done)) {
    await command.execute()
  }
})()

const dispatch = async (command: ListCommand) => {
  history.value.push(command)
  await command.execute()
  return command
}

export const useCommands = () => {
  return {
    dispatch,
    history,
  }
}

export const mediaListToForm = (
  mediaListEntry: MediaList | null,
  advancedScoring: string[],
): Form => {
  if (mediaListEntry) {
    const advancedScores = advancedScoring
      .map(key => mediaListEntry.advancedScores[key])
      .map(value => value || 0)

    delete mediaListEntry.__typename
    delete mediaListEntry.startedAt.__typename
    delete mediaListEntry.completedAt.__typename
    delete mediaListEntry.private
    delete mediaListEntry.id
    delete mediaListEntry.mediaId

    return {
      ...mediaListEntry,
      advancedScores,
    }
  }

  const advancedScores = advancedScoring.map(() => 0)
  return {
    advancedScores,
    completedAt: { day: null, month: null, year: null },
    notes: '',
    progress: 0,
    progressVolumes: 0,
    repeat: 0,
    score: 0,
    startedAt: { day: null, month: null, year: null },
    status: null,
  }
}

export const isDate = (e: any): e is FuzzyDate => e.__typename === 'FuzzyDate'
