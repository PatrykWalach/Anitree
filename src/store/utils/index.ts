export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
