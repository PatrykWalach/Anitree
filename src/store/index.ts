import Vue from 'vue'
import Vuex from 'vuex'

// import { getModule } from 'vuex-module-decorators'
Vue.use(Vuex)

import { ModuleMedia } from './modules/media'
import { ModuleTheme } from './modules/theme'
import { ModuleTitle } from './modules/title'
import { ModuleAuth } from './modules/auth'

export interface RootState {
  media: ModuleMedia
  theme: ModuleTheme
  title: ModuleTitle
  auth: ModuleAuth
}

export default new Vuex.Store<RootState>({})
