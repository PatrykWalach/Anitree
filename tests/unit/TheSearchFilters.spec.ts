import TheSearchFilters from '@/components/TheSearchFilters.vue'
import { matchSnapshot } from './utils'
import { useMockedStore } from './mocks/store'

describe('TheSearchFilters', () => {
  matchSnapshot(TheSearchFilters, {
    mocks: {
      $route: {
        query: {},
      },
    },
    ...useMockedStore(),
  })
})
