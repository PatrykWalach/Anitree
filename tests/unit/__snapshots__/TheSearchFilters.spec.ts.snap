// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheSearchFilters matches snapshot 1`] = `
<v-dialog-stub
  closedelay="0"
  contentclass=""
  maxwidth="360px"
  opendelay="0"
  origin="center center"
  retainfocus="true"
  scrollable="true"
  transition="dialog-transition"
  width="auto"
>
  <v-card-stub
    loaderheight="4"
    tag="div"
  >
    <v-card-title-stub>
      
      Filters
    
    </v-card-title-stub>
     
    <v-card-text-stub
      class="pa-0"
    >
      <v-card-text-stub>
        <v-row-stub
          tag="div"
        >
          <v-col-stub
            cols="12"
            tag="div"
          >
            <span
              class="subheading"
            >
              Type
            </span>
             
            <v-chip-group
              active-class="accent--text"
              column=""
            >
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="ANIME"
              >
                
                anime
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="MANGA"
              >
                
                manga
              
              </v-chip-stub>
            </v-chip-group>
          </v-col-stub>
           
          <v-col-stub
            cols="12"
            tag="div"
          >
            <span
              class="subheading"
            >
              Sort
            </span>
             
            <v-chip-group
              active-class="accent--text"
              column=""
            >
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="ID_DESC"
              >
                
                Date Added
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="FAVOURITES_DESC"
              >
                
                Favorites
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="POPULARITY_DESC"
              >
                
                Popularity
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="START_DATE_DESC"
              >
                
                Release Date
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="SCORE_DESC"
              >
                
                Score
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="TITLE_ROMAJI"
              >
                
                Title
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filtericon="$complete"
                tag="span"
                value="TRENDING_DESC"
              >
                
                Trending
              
              </v-chip-stub>
            </v-chip-group>
          </v-col-stub>
           
          <v-col-stub
            cols="12"
            tag="div"
          >
            <span
              class="subheading"
            >
              Status
            </span>
             
            <base-field-stub
              active-class="accent--text"
              aftertransform="function (e) { return (e.length ? e : undefined); },function (status) {
                  _vm.form = { status: status }
                }"
              beforetransform="function (e) { return (e ? (e instanceof Array ? e : [e]) : []); }"
              column=""
              multiple=""
              tag="v-chip-group"
              transformations=""
              validators=""
              value=""
            >
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filter="true"
                filtericon="$complete"
                tag="span"
                value="CANCELLED"
              >
                
                cancelled
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filter="true"
                filtericon="$complete"
                tag="span"
                value="FINISHED"
              >
                
                finished
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filter="true"
                filtericon="$complete"
                tag="span"
                value="NOT_YET_RELEASED"
              >
                
                not yet released
              
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="text-capitalize"
                closeicon="$delete"
                filter="true"
                filtericon="$complete"
                tag="span"
                value="RELEASING"
              >
                
                releasing
              
              </v-chip-stub>
            </base-field-stub>
          </v-col-stub>
        </v-row-stub>
      </v-card-text-stub>
       
      <v-list-stub
        flat="true"
        subheader="true"
        tag="div"
        tile="true"
      >
        <v-subheader-stub>
          Other
        </v-subheader-stub>
         
        <v-list-item-group-stub
          activeclass="v-item--active"
        >
          <v-list-item-stub
            activeclass=""
            tag="div"
            value="false"
          />
           
          <v-list-item-stub
            activeclass=""
            tag="div"
            value="true"
          />
        </v-list-item-group-stub>
         
        <v-list-item-group-stub
          activeclass="v-item--active"
        >
          <v-list-item-stub
            activeclass=""
            tag="div"
            value="true"
          />
        </v-list-item-group-stub>
      </v-list-stub>
    </v-card-text-stub>
     
    <v-divider-stub />
     
    <v-card-actions-stub>
      <v-spacer-stub />
       
      <v-btn-stub
        activeclass=""
        color="error"
        tag="button"
        text="true"
        type="button"
      >
        cancel
      </v-btn-stub>
       
      <v-btn-stub
        activeclass=""
        color="primary"
        disabled="true"
        tag="button"
        text="true"
        type="button"
      >
        submit
      </v-btn-stub>
    </v-card-actions-stub>
  </v-card-stub>
</v-dialog-stub>
`;
